### ISSUE TITLE
Sometimes multiplayer shows up empty.

### ISSUE THREAD

**Issue by rtwfroody** [ZeroK-RTS/Chobby#776]
E.g.:
![image](https://user-images.githubusercontent.com/3102687/99328909-e5de6d00-2831-11eb-960b-29f142c5dfa2.png)

When that happens, there's always some JSON error in the log. This time:
[f=-000001] [liblobby] Error: Failed to parse JSON: {"Header":{"BattleID":1256,"Engine":"104.0.1-1544-ge1f249f","Founder":"DroneZone","Game":"Zero-K v1.8.11.2","IsMatchMaker":false,"IsRunning":true,"Map":"Hide_and_Seek_v03","MaxPlayers":10,"Mode":5,"Password":"?","PlayerCount":3,"RunningSince":"2020-11-17T00:24:27.996582Z","SpectatoWelcome {"Engine":"104.0.1-1544-ge1f249f","Game":"Zero-K v1.8.11.2","UserCount":98,"Version":"1.4.9.246","Blacklist":["dakeys","xtrasauce"],"Factions":[{"Name":"Dynasty of Man","Shortcut":"Dynasty","Color":"#FFBF00"},{"Name":"Empire","Shortcut":"Empire","Color":"#6010FF"},{"Name":"Federation","Shortcut":"Federation","Color":"#55BB55"}],"UserCountLimited":false}

Full log:
[infolog.txt](https://github.com/ZeroK-RTS/Chobby/files/5550692/infolog.txt)

(Originally filed at https://github.com/ZeroK-RTS/Zero-K/issues/4194)

**Comment by rtwfroody** [comment]
If you look closely, you see that the JSON being parsed seems to be missing something, or has overwritten something:

{"Header":{"BattleID":1256,"Engine":"104.0.1-1544-ge1f249f","Founder":"DroneZone","Game":"Zero-K v1.8.11.2","IsMatchMaker":false,"IsRunning":true,"Map":"Hide_and_Seek_v03","MaxPlayers":10,"Mode":5,"Password":"?","PlayerCount":3,"RunningSince":"2020-11-17T00:24:27.996582Z","**SpectatoWelcome** {"Engine":"104.0.1-1544-ge1f249f","Game":"Zero-K v1.8.11.2","UserCount":98,"Version":"1.4.9.246","Blacklist":["dakeys","xtrasauce"],"Factions":[{"Name":"Dynasty of Man","Shortcut":"Dynasty","Color":"#FFBF00"},{"Name":"Empire","Shortcut":"Empire","Color":"#6010FF"},{"Name":"Federation","Shortcut":"Federation","Color":"#55BB55"}],"UserCountLimited":false}

I am currently experiencing the error with this JSON:
{"NewsItems":[{"Header":"Zero-K 1.8.11.3 - Detriment and Venom tweaks","Text":"This update addresses the slightly overpowered stats on the reworked Venom. While we're at it, here are some improvements and fixes for Detriment. There is also a tournament this weekend which is open for anyone to join or watch.","Url":"https://zero-k.info/Forum/Thread/32470","Image":"http://zero-k.info/img/lobbynews/1097.jpg"},{"Header":"1v1 Tournament November 21st 2020","Text":"There will be a 1v1 tournament at 19:00 UTC on Saturday November 21st - Click on the link to see the tournament thread and sign up!","Time":"2020-11-21T19:00:00","Url":"http://zero-k.info/Forum/Thread/32437","Image":"http://zero-k.info/img/lobbynews/1096.png"},{"Header":"Zero-K v1.8.11.1 - Bulkhead, Balancing and Mod Hosting","Text":"The Amphibious Factory now has a full roster with the addition of Bulkhead - a deployable fire support unit with impressive health and range. This update also includes a wide range of balance changes, from 3% more DPS for Glaive to 200% more awesomeness for Detriment. Additionally, hosting mods is now easier than ever and there are a few other fixes and features.\r\n\r\nThe balance changes focus on improving the specialist factories (as was decided by a poll on Patreon). Apart from Bulkhead solving Amph's pre-Grizzly lack of range, Venom is more reliable and a bug that makes jumpjets hard to tune was fixed. Other factories were also tweaked and Amph received a few extra changes to accommodate Bulkhead and the Archer rework.","Url":"http://zero-k.info/Forum/Thread/32417","Image":"http://zero-k.info/img/lobbynews/1095.jpg"},{"Header":"Zero-K v1.8.10.0 - Eight New Maps and MM Rotation","Text":"Aquanim and Dregs have put together a mammoth effort and released eight new maps. Dregs made  Desert Rumble, Prestige Arena, Mechadansonia, Jurassic Sands, Sparkles Reef and Scaryland, while Aquanim made Anvilwood and Cobalt Dream. \r\n\r\nThese maps form the basis of a major matchmaker rotation, but are also suitable for coop, small teams, and possibly even arena-style large games. The update also has improved explosion scars, an experimental new AI to test and improved mod support.","Url":"http://zero-k.info/Forum/Thread/32255","Image":"http://zero-k.info/img/lobbynews/1094.jpg"},{"Header":"Zero-K v1.8.9.4 - Fixes and Amph Utility","Text":"Hotfixes don't usually warrant a full set of notes, but this one contains a few things that were being worked on in the meantime and fixes a noticeable idle behaviour bug. We're preparing to take a look at some of the less well-rounded factories, and **part Welcome {"Engine**":"104.0.1-1544-ge1f249f","Game":"Zero-K v1.8.11.3","UserCount":107,"Version":"1.4.9.248","Blacklist":["dakeys","xtrasauce"],"Factions":[{"Name":"Dynasty of Man","Shortcut":"Dynasty","Color":"#FFBF00"},{"Name":"Empire","Shortcut":"Empire","Color":"#6010FF"},{"Name":"Federation","Shortcut":"Federation","Color":"#55BB55"}],"UserCountLimited":false}

Looks like again something clobbered a buffer maybe.

**Comment by rtwfroody** [comment]
Here's another. It seems there's always the Welcome message that messes up the parsing:
{"Header":{"BattleID":2101,"Engine":"104.0.1-1544-ge1f249f","Founder":"YCCCM7","Game":"Zero-K v1.8.11.3","IsMatchMaker":false,"IsRunning":true,"Map":"FrozenFortress_v2","MaxPlayers":10,"Mode":0,"Password":"?","PlayerCount":1,"RunningSince":"2020-11-26T00:13:16.0172708Z","SpectatorCount":0,"Title":"YCCCM7's **BatWelcome** {"Engine":"104.0.1-1544-ge1f249f","Game":"Zero-K v1.8.11.3","UserCount":71,"Version":"1.4.9.248","Blacklist":["dakeys","xtrasauce"],"Factions":[{"Name":"Dynasty of Man","Shortcut":"Dynasty","Color":"#FFBF00"},{"Name":"Empire","Shortcut":"Empire","Color":"#6010FF"},{"Name":"Federation","Shortcut":"Federation","Color":"#55BB55"}],"UserCountLimited":false}

**Comment by GoogleFrog** [comment]
Sounds like an infrastructure problem.

**Comment by Licho1** [comment]
Does it? Its using standard serializer is the json generated invalid?

**Comment by GoogleFrog** [comment]
The json is invalid.

**Comment by Licho1** [comment]
This is extremely odd. Basically some letters are skipped" How could this even happen?
Did it happen to anyone else ever? If it was corrupted buffer it would do something else. It looks like wrong "stiching" of messages. 

Check the code that waits for entire line/command to be recieved before processing it. 
Server sends each command/line as a single operating system operation (filling up buffer) so it's very unlikely to be problem with outgoing data.
Rather it has to be stiching problem in chobby.

**Comment by Licho1** [comment]
Check interface_shared.lua:_SocketUpdate() 

It's not doing any stiching apparently!  What can happen in practice is that you call input:receive and it gives you incomplete command.
You always need to wait for \n before processing it!

When and how often it happens depend on network configuration (max packet size etc).

**Comment by GoogleFrog** [comment]
Is there a way for me to test this?

**Comment by Licho1** [comment]
Generate some giant lobby message. Perhaps oversized modoptions would do?

**Comment by rtwfroody** [comment]
It looks like the from the server is good. I managed to collect a packet log with wireshark.

First, here's the error in this particular instance:
[f=-000001] [liblobby] Error: Failed to parse JSON: {"NewsItems":[{"Header":"Zero-K 1.8.11.3 - Detriment and Venom tweaks","Text":"This update addresses the slightly overpowered stats on the reworked Venom. While we're at it, here are some improvements and fixes for Detriment. There is also a tournament this weekend which is open for anyone to join or watch.","Url":"https://zero-k.info/Forum/Thread/32470","Image":"http://zero-k.info/img/lobbynews/1097.jpg"},{"Header":"1v1 Tournament November 21st 2020","Text":"There will be a 1v1 tournament at 19:00 UTC on Saturday November 21st - Click on the link to see the tournament thread and sign up!","Time":"2020-11-21T19:00:00","Url":"http://zero-k.info/Forum/Thread/32437","Image":"http://zero-k.info/img/lobbynews/1096.png"},{"Header":"Zero-K v1.8.11.1 - Bulkhead, Balancing and Mod Hosting","Text":"The Amphibious Factory now has a full roster with the addition of Bulkhead - a deployable fire support unit with impressive health and range. This update also includes a wide range of balance changes, from 3% more DPS for Glaive to 200% more awesomeness for Detriment. Additionally, hosting mods is now easier than ever and there are a few other fixes and features.\r\n\r\nThe balance changes focus on improving the specialist factories (as was decided by a poll on Patreon). Apart from Bulkhead solving Amph's pre-Grizzly lack of range, Venom is more reliable and a bug that makes jumpjets hard to tune was fixed. Other factories were also tweaked and Amph received a few extra changes to accommodate Bulkhead and the Archer rework.","Url":"http://zero-k.info/Forum/Thread/32417","Image":"http://zero-k.info/img/lobbynews/1095.jpg"},{"Header":"Zero-K v1.8.10.0 - Eight New Maps and MM Rotation","Text":"Aquanim and Dregs have put together a mammoth effort and released eight new maps. Dregs made  Desert Rumble, Prestige Arena, Mechadansonia, Jurassic Sands, Sparkles Reef and Scaryland, while Aquanim made Anvilwood and Cobalt Dream. \r\n\r\nThese maps form the basis of a major matchmaker rotation, but are also suitable for coop, small teams, and possibly even arena-style large games. The update also has improved explosion scars, an experimental new AI to test and improved mod support.","Url":"http://zero-k.info/Forum/Thread/32255","Image":"http://zero-k.info/img/lobbynews/1094.jpg"},{"Header":"Zero-K v1.8.9.4 - Fixes and Amph Utility","Text":"Hotfixes don't usually warrant a full set of notes, but this one contains a few things that were being worked on in the meantime and fixes a noticeable idle behaviour bug. We're preparing to take a look at some of the less well-rounded factories, and part of that is making their unique aspects feel better to use. A few improvements to Lobster and Djinn made it into this update. Most notable is Djinn's less restrictive requirement for teleport.","Url":"https://zero-k.info/Forum/Thread/32121","Image":"http://zero-k.info/img/lobbynews/1093.png"},{"Header":"Zero-K v1.8.9.1 - Control Cleanup and Idle Improvement","Text":"Units are smarter about chasing enemies long distances (they don't), the command panel is clearer, and there are many other interface tweaks and improvements. The changes were aimed at making the game more approachable for new players, but the cleaner UI is nice enough for everyone to benefit. If you prefer more complexity in a particular part of the interface then read the main thread to discover how to configure it to your liking.","Url":"https://zero-k.info/Forum/Thread/32102","Image":"http://zero-k.info/img/lobbynews/1092.jpg"},{"Header":"Zero-K v1.8.8.2 - Construction Plates and Archer Rework","Text":"The big feature of this update is the introduction of construction plates - efficient mini-factories that must be placed near a main factory to function. They can be used in the campaign and teamgames to take advantage of allied tech, while in 1v1 they should make it easier for factories access their heavy units.\r\n\r\nOther changes include Kodachi nerfs to account for fire prediction, improved Amph interaction with sea, and nerfs for Funnelweb, Lance, *GWelcome* {"Engine":"104.0.1-1544-ge1f249f","Game":"Zero-K v1.8.11.3","UserCount":81,"Version":"1.4.9.248","Blacklist":["dakeys","xtrasauce"],"Factions":[{"Name":"Dynasty of Man","Shortcut":"Dynasty","Color":"#FFBF00"},{"Name":"Empire","Shortcut":"Empire","Color":"#6010FF"},{"Name":"Federation","Shortcut":"Federation","Color":"#55BB55"}],"UserCountLimited":false}

Here's the packet log, which you can open in wireshark:
[Zero-K-JSON-bug.pcapng.gz](https://github.com/ZeroK-RTS/Chobby/files/5626229/Zero-K-JSON-bug.pcapng.gz)

Then look at packet 261. Wireshark gives you the option to view the "Reassembled TCP (121970 bytes)". In there, you can see the message that was cut off in the error message existing completely:
...and nerfs for Funnelweb, Lance, Gauss and Big Bertha to open up the options in dense team games.</a:Text>...

The Welcome message that overwrites this buffer happens later in the packet stream, at packet 326.

**Comment by Licho1** [comment]
Yeah there is missing stiching code in lobby we just dont have super easy way to test it / simulate huge packets. We might try to create tester code .. @GoogleFrog  did you manage to cause split with giant mod options?  Should I make you a test program?

**Comment by GoogleFrog** [comment]
I have not tried yet. A test program would be good.

**Comment by GoogleFrog** [comment]
It appears to be stitching.

```local commands = explode("\n", commandsStr)
			commands[1] = self.buffer .. commands[1]
			for i = 1, #commands-1 do
				local command = commands[i]
				if command ~= nil then
					self:CommandReceived(command)
				end
			end
			self.buffer = commands[#commands]
```

**Comment by GoogleFrog** [comment]
I caught an example of it stitching correctly.
```
[t=00:03:57.050777][f=0003905] self.buffer, BattleAdded {"Header":{"BattleID":4681,"Engine":"104.0.1-1544-ge1f249f","Founder":"ecogreen123","Game":"Zero-K v1.8.11.3","IsMatchMaker":false,"IsRunning":false,"Map":"Rogues River v1.2","MaxPlayers":16,"Mode":4,"Password":"?","PlayerCount":1,"SpectatorCount":0,"Title":"ecogreen123's Battle","TimeQueueEnabled":true,"MaxEvenPlayer
[t=00:03:57.051777][f=0003905] commandsStr, s":0}}
```

**Comment by GoogleFrog** [comment]
Press this button if it happens again https://github.com/ZeroK-RTS/Chobby/commit/8ead7f8ae29f190b1eb5399959f3dae9842b305c

**Comment by Licho1** [comment]
note this code is only correct if explode("something\n") returns array with two values.

**Comment by Licho1** [comment]
This is really odd then, could it be some processing madness with ```\r\n``` vs ```\n```?

**Comment by GoogleFrog** [comment]
Yes, explode("something\n") = {"something", ""}. I guess \r could be a factor, but I've never seen it.

**Comment by rtwfroody** [comment]
> Press this button if it happens again 8ead7f8

How do I get to see the button? There appears to be something called "dev mode" where you get a Settings->Developer menu, but I don't have that. I don't easily see a launch option for it in Steam either.

**Comment by GoogleFrog** [comment]
It isn't out yet, although you could also start zk with zkmenu:test to try it out early. The wiki explains devmode.

**Comment by rtwfroody** [comment]
Now with "Debug server messages" and "Debug raw server messages" both enabled:
```
[f=-000001] [liblobby] Error: Failed to parse JSON: {"Header":{"BattleID":4864,"Engine":"104.0.1-1544-ge1f249f","FoundeWelcome {"Engine":"104.0.1-1544-ge1f249f","Game":"Zero-K v1.8.12.0"
,"UserCount":109,"Version":"1.4.9.249","Blacklist":["dakeys","xtrasauce"],"Factions":[{"Name":"Dynasty of Man","Shortcut":"Dynasty","Color":"#FFBF00"},{"Name":"Empire","Shortcut":"Empire
","Color":"#6010FF"},{"Name":"Federation","Shortcut":"Federation","Color":"#55BB55"}],"UserCountLimited":false}
[f=-000001] [liblobby] Error: [string "libs/liblobby/lobby/interface_zerok.lua"]:1023: attempt to index local 'header' (a nil value)
[f=-000001] [liblobby] Error: [string "libs/liblobby/lobby/interface_zerok.lua"]:1023: attempt to index local 'header' (a nil value)
stack traceback:
        [string "libs/liblobby/lobby/observable.lua"]:51: in function '_PrintError'
        [string "libs/liblobby/lobby/interface_shared.lua"]:282: in function <[string "libs/liblobby/lobby/interface_shared.lua"]:282>
        [string "libs/liblobby/lobby/interface_zerok.lua"]:1023: in function 'jsonCommandFunction'
        [string "libs/liblobby/lobby/interface_shared.lua"]:199: in function '_OnCommandReceived'
        [string "libs/liblobby/lobby/interface_shared.lua"]:149: in function 'CommandReceived'
        [string "libs/liblobby/lobby/interface_shared.lua"]:252: in function '_SocketUpdate'
        [string "libs/liblobby/lobby/interface_shared.lua"]:270: in function 'SafeUpdate'
        [string "libs/liblobby/lobby/interface_shared.lua"]:281: in function <[string "libs/liblobby/lobby/interface_shared.lua"]:281>
        [C]: in function 'xpcall'
        [string "libs/liblobby/lobby/interface_shared.lua"]:281: in function 'Update'
        [string "libs/liblobby/LuaMenu/widgets/api_lobby.lua"]:70: in function <[string "libs/liblobby/LuaMenu/widgets/api_lobby.lua"]:69>
        (tail call): ?
        [C]: in function 'pcall'
        [string "LuaHandler/Utilities/crashHandler.lua"]:50: in function 'f'
        [string "LuaHandler/Utilities/specialCallinHandlers...."]:80: in function <[string "LuaHandler/Utilities/specialCallinHandlers...."]:76>
```

Full log: 
[infolog.txt](https://github.com/ZeroK-RTS/Chobby/files/5700167/infolog.txt)

**Comment by rtwfroody** [comment]
Here's another one, in case more is better: 
[infolog.txt](https://github.com/ZeroK-RTS/Chobby/files/5706584/infolog.txt)
